R/utils.R

```diff
@@ -407,4 +407,58 @@ findColNamesByPredicate <- function(variableList, predicateFunction) {
 
 avgDigits <- function(x) {
   floor(mean(stringi::stri_count_regex(as.character(x), "[[:digit:]]")))
+}
+
+
+#
+# For any number, return an absolute delta (numeric) at the last
+# significant digit in the number, using the number of digits specified
+#
+# e.g. assuming 3 significant digits
+# 
+# 1.23 -> 0.01
+# 11.0 -> 0.1
+# 12.3 -> 0.1
+# 101000 -> 1000
+# 1.20e-05 -> 0.01e-05 == 1.0e-07
+# 0.0123e-05 -> 0.0001e-05 == 1.0e-09
+# -2.34e-02 -> 0.01e-02 == 1.0e-04
+# 
+signifDigitDelta <- function(x, digits) {
+
+  # '#' flag ensure trailing zeroes
+  # take abs() here because we don't care about sign
+  rounded <- formatC(abs(x), digits = digits, width = 1L, flag = '#')
+
+  # split into vector of single characters
+  characters <- strsplit(rounded, '')
+
+  result <- c()
+  seenSignificant <- FALSE
+  significantCount <- 0
+  # walk through string, looking for first non-zero, non decimal point character
+  for (c in unlist(characters)) {
+    if (!(c %in% c('0', '.'))) {
+      seenSignificant <- TRUE
+    }
+    if (c == '.') {
+      result <- c(result, c)
+    } else if (seenSignificant) {
+      significantCount <- significantCount + 1
+      if (significantCount == digits) {
```

Comment:
I'll put the `significantCount < digits` condition first, before `== digits` so that it is easier to understand what's going on.

Reply:
Done